apps/aecontract/src/aect_state_tree.erl

```diff
@@ -75,37 +68,28 @@ lookup_contract(Id, Tree) ->
 %% -- Calls --
 
 -spec insert_call(aect_call:call(), tree()) -> tree().
-insert_call(Call, Tree = #contract_tree{ contracts = CtTree, calls = CallTrees }) ->
-    %% Get the contract being called, and its call history
+insert_call(Call, Tree = #contract_tree{ contracts = CtTree}) ->
+    %% Construct the Id to store in the tree.
     CtId       = aect_call:contract_address(Call),
-    Contract   = get_contract(CtId, Tree),
-    CallTree   = gb_trees:get(CtId, CallTrees),
-
-    %% Insert the new call into the history
     CallId     = aect_call:id(Call),
-    CallTree1  = aeu_mtrees:insert(CallId, aect_call:serialize(Call), CallTree),
+    CallTreeId = <<CtId/binary, CallId/binary>>,
 
-    %% Update the call tree hash in the contract
-    {ok, Hash} = aeu_mtrees:root_hash(CallTree1),
-    Contract1  = aect_contracts:set_calls_hash(Hash, Contract),
-
-    %% Update the contract tree and the calls tree
-    CtTree1    = aeu_mtrees:enter(CtId, aect_contracts:serialize(Contract1), CtTree),
-    CallTrees1 = gb_trees:enter(CtId, CallTree1, CallTrees),
+    %% Insert the new call into the history
+    Serialized = aect_call:serialize(Call),
+    CtTree1    = aeu_mtrees:insert(CallTreeId, Serialized, CtTree),
```

Comment:
This means inserting in tree both contracts and calls.
This also means inserting in same tree with key of different length.

(I am trying to understand rationale behind and impact.)

Reply:
The rational is to keep all things in the MPTree so that it will get the befefits as other tree nodes (primarily that nodes are shared between block states).

Earlier, the forrest of gb_trees was getting duplicated for all block states.

Prepending the parent id is for finding the calls to one particular contract. This is not used here yet (See oracles)